src/java/org/apache/cassandra/gms/EndpointState.java

```diff
@@ -114,6 +128,47 @@ public void addApplicationStates(Set<Map.Entry<ApplicationState, VersionedValue>
         }
     }
 
+    void removeMajorVersion3LegacyApplicationStates()
+    {
+        while (hasLegacyFields())
+        {
+            Map<ApplicationState, VersionedValue> orig = applicationState.get();
+            Map<ApplicationState, VersionedValue> updatedStates = filterMajorVersion3LegacyApplicationStates(orig);
+            // avoid updating if no state is removed
+            if (orig.size() == updatedStates.size()
+                || applicationState.compareAndSet(orig, updatedStates))
+                return;
+        }
+    }
+
+    private boolean hasLegacyFields()
+    {
+        Set<ApplicationState> statesPresent = applicationState.get().keySet();
+        if (statesPresent.isEmpty())
+            return false;
+        return (statesPresent.contains(STATUS) && statesPresent.contains(STATUS_WITH_PORT))
+               || (statesPresent.contains(INTERNAL_IP) && statesPresent.contains(INTERNAL_ADDRESS_AND_PORT))
+               || (statesPresent.contains(RPC_ADDRESS) && statesPresent.contains(NATIVE_ADDRESS_AND_PORT));
+    }
+
+    private Map<ApplicationState, VersionedValue> filterMajorVersion3LegacyApplicationStates(Map<ApplicationState, VersionedValue> states)
```

Comment:
```suggestion
    private static Map<ApplicationState, VersionedValue> filterMajorVersion3LegacyApplicationStates(Map<ApplicationState, VersionedValue> states)
```

Reply:
Is the motivation of making it `static` to not pass `this` reference and not able to access instance fields? make sense...

Reply:
yes. for the reader it's a optimisation: the following code has nothing to do with any instance fields. nothing more than that here.